vtqueryserver

demmer opened the pull request:
# What 

Add a new binary called `vtqueryserver` which couples a standalone tabletserver with a grpc interface and a simple mysql server implementation. This can be deployed alongside any mysql server to provide connection pooling, query consolidation, logging, and other features of vttablet, without any query rewriting, topology management, replication control, or the need for a vtgate tier for query routing.

# Why
Although our intent is to migrate everything to a new vitess cluster, our existing mysql workload would benefit from a less-intrusive deployment in which we can still get some of the performance benefits of the tablet on our existing systems.

By skipping the query rewriting and routing elements of vitess, adding this proxy to an existing system is a lower risk change yet still provides some of the performance benefits. Also since the `vtqueryserver` supports both the mysql protocol and a GRPC interface (similar to vtgate) it can also enable long-lived connections and splitting of queries from bind variables.

sougou commented:
Change looks good overall. Some changes needed:
* Conflicts to be reseoved
* Tests need fixing
* mysql plugin autocommit behavior needs to be fixed.

demmer commented:
@sougou I reworked the autocommit behavior and added an endtoend test for vtqueryserver including test cases for passthrough DML support.

If you could please take a look and make sure my understanding of mysql autocommit behavior is correct.

I still need to do another round of testing with the actual binary before I'm sure this is ready to merge -- will update here when that's done.

sougou commented on go/vt/vtqueryserver/endtoend_test.go:
```diff
@@ -0,0 +1,314 @@
+	}
+}
+
+func TestAutocommitDMLs(t *testing.T) {
```
I couldn't find a test for autocommit=0 -> autocommit=1 transition (where it needs to commit the current transaction). Ref: https://github.com/youtube/vitess/blob/master/go/vt/vtgate/executor_test.go#L348

sougou commented on go/vt/vtqueryserver/endtoend_test.go:
```diff
@@ -0,0 +1,314 @@
+	testFetch(t, conn2, "select * from test", 0)
+}
+
+func TestTransactions(t *testing.T) {
```
I couldn't see a test where a begin->dmls->begin causes the previous transaction to be commited and a new one started.

sougou commented on go/vt/mysqlproxy/mysqlproxy.go:
```diff
@@ -0,0 +1,226 @@
+}
+
+func (mp *Proxy) doBegin(ctx context.Context, session *ProxySession) error {
+	txID, err := mp.qs.Begin(ctx, mp.target, session.Options)
```
This needs to check if we are in a transaction. If so, it must be committed. Ref: https://github.com/youtube/vitess/blob/master/go/vt/vtgate/tx_conn.go#L52.

sougou commented:
I've reviewed the rest of the code. Once the new test cases are added, this is good to go.

demmer commented:
@sougou I fixed the issue you pointed out and added tests for the two missing cases.

sougou commented:
LGTM
